nixos-rebuild: Added a version check

crazazy opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Before this patch, if /etc/nixos/flake.nix existed, nixos-rebuild would
try to use flakes even if you had an unsupported version of nix. we now
do an additional version check to see if nix can use flakes in the first
place.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

crazazy committed:
Added a version check to nixos rebuild

Before this patch, if /etc/nixos/flake.nix existed, nixos-rebuild would
try to use flakes even if you had an unsupported version of nix. we now
do an additional version check to see if nix can use flakes in the first
place.

Synthetica9 commented:
This is gonna cause a problem for nix 2.12.1-3 🤔

r-rmcgibbo commented:
Result of `nixpkgs-review pr 117707` at 1594d240 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>nixos-rebuild</li>
  </ul>
</details>

crazazy committed:
fixed some issue with semi-colons

crazazy commented:
r

> This is gonna cause a problem for nix 2.12.1-3 thinking

As far as I am aware, after this we should be moving to 3.0
thought I might want to add another dot to the regex so that we dont accidentally check for a minor+hotfix combination yeah...

crazazy committed:
make sure that we are looking for a major+minor version

Previously, nixos-rebuild would mistakenly assume an old version in the
case nix hits 3.2.{1,2,3}
